tlsutil/config.go

```diff
@@ -36,13 +36,42 @@ type DCWrapper func(dc string, conn net.Conn) (net.Conn, error)
 // a constant value. This is usually done by currying DCWrapper.
 type Wrapper func(conn net.Conn) (net.Conn, error)
 
-// tlsLookup maps the tls_min_version configuration to the internal value
-var tlsLookup = map[string]uint16{
-	"":      tls.VersionTLS10, // default in golang
-	"tls10": tls.VersionTLS10,
-	"tls11": tls.VersionTLS11,
-	"tls12": tls.VersionTLS12,
-	"tls13": tls.VersionTLS13,
+// goTLSVersions maps types.TLSVersion to the Go internal value
+var goTLSVersions = map[types.TLSVersion]uint16{
+	types.TLSVersionAuto: tls.VersionTLS10, // default in golang
+	types.TLSv1_0:        tls.VersionTLS10,
+	types.TLSv1_1:        tls.VersionTLS11,
+	types.TLSv1_2:        tls.VersionTLS12,
+	types.TLSv1_3:        tls.VersionTLS13,
+}
+
+// ParseTLSVersion maps a TLS version configuration string to the internal type
+func ParseTLSVersion(tlsVersionString string) (types.TLSVersion, error) {
+	// Handle empty string case for unspecified config.
+	//
+	// This check is not inside types.ValidateTLSVersionString because Envoy config
+	// distinguishes between an unset empty string which inherits parent config and
+	// an explicit TLS_AUTO which allows overriding parent config with the proxy
+	// defaults.
+	if tlsVersionString == "" {
+		return types.TLSVersionAuto, nil
+	}
+
+	v := types.TLSVersion(tlsVersionString)
+
+	tlsVersionErr := types.ValidateTLSVersion(v)
+	if tlsVersionErr != nil {
+		// NOTE: This inner check for deprecated values should eventually be
+		// removed
+		if v, ok := types.DeprecatedConsulAgentTLSVersions[tlsVersionString]; ok {
+			// TODO: log warning about deprecated config
+			return v, nil
+		}
+
+		// Only suggest non-deprecated values if configured value is invalid
+		return types.TLSVersionInvalid, tlsVersionErr
+	}
+	return v, nil
```

Comment:
What do you think about moving this to the config package, and handling the deprecated values in `applyDeprecatedTLSConfig`? I think it's best to try to keep as much of the complexity of handling previous configuration formats in one place so that they don't leak into the rest of the codebase.

Reply:
Yea, I think that makes sense - this function isn't called anywhere else, and the shared logic now lives in https://github.com/hashicorp/consul/blob/a1110845df0e4b1dfb588c4577484029b41bb01f/types/tls.go#L95-L101